src/rust/lib/graph/src/component/node/port.rs

```diff
@@ -0,0 +1,319 @@
+
+        let triangle = Triangle(&tri_width, &tri_height);
+        let triangle = triangle.rotate(180.0.deg().radians());
+        let tri_offset: Var<Distance<Pixels>> = tri_height.clone().into();
+        let triangle = triangle.translate_y(tri_offset);
+        // TODO[mm] consider replace with a `Plane().cut_angle`
+        // But avoid visual artifacts at the other end of the circle.
+        // let section = Plane().cut_angle(&spec.width);
+        // let section = section.rotate(180.0.deg().radians());
+        // let section = section.translate_y(tri_offset);
+
+        let circle_radius: Var<Distance<Pixels>> = outer_radius.clone().into();
+        let circle_outer    = Circle((circle_radius));
+
+        let circle_offset_y: Var<Distance<Pixels>> = (&tri_height - &outer_radius).into();
+        let circle_outer    = circle_outer.translate_y((circle_offset_y));
+
+        let triangle_rounded = Intersection(triangle,circle_outer);
+        let triangle_rounded = triangle_rounded.fill(Srgb::new(0.26, 0.69, 0.99));
+
+        triangle_rounded.into()
+    }
+
+    /// Canvas node shape definition.
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32) {
+            // TODO[mm] take spec or spec values as `Var<_>` parameters
+            let port_spec_val = SpecificationVar{
+                    height       : height.into(),
+                    width        : width.into(),
+                    inner_radius : inner_radius.into()
+            };
+            new_port_inwards(port_spec_val)
+        }
+    }
+
+}
+
+
+mod shape_out{
+    use super::*;
+    use ensogl::display::shape::*;
+    use ensogl::display::shape::primitive::def::modifier::immutable::*;
+    use ensogl::display::shape::primitive::def::primitive::*;
+    use ensogl::prelude::*;
+    use ensogl::display::shape::primitive::def::class::ShapeOps;
+    use ensogl::math::geometry::circle::circle_segment::CircleSegment;
+    use ensogl::math::topology::unit::AngleOps;
+    use ensogl::math::topology::unit::Distance;
+    use ensogl::math::topology::unit::PixelDistance;
+    use ensogl::math::topology::unit::Pixels;
+    use nalgebra as na;
+
+    /// Construct an outwards facing port.
+    fn new_port_outwards(spec:SpecificationVar) -> AnyShape {
+        // TODO[mm] cut down on clone usage
+
+        let inner_radius : Var<f32> = spec.inner_radius.into();
+        let height       : Var<f32> = spec.height.clone().into();
+
+        let segment_width_rad: Var<f32>       = spec.width.clone().into();
+        let segment_radius   : Var<f32>       = inner_radius.clone().into();
+        let segment : CircleSegment<Var<f32>> = CircleSegment::new(segment_radius,segment_width_rad);
+
+        let tri_base             = segment.sagitta();
+        let tri_height: Var<f32> = &height + &tri_base;
+        let tri_width            = segment.chord_length();
+
```

Comment:
I was already writing about it. I don't like that input / output ports are separate shapes. Why this design? Why they are not a single shape but parametrized?

Reply:
The way the shapes are currently constructed, one of them needs to be a `Union` between a `Circle` and a `Triangle` and the other one a `Difference`. 
I think the only way to make this work, might be to use additional shapes like always having one `Circle` that computes the difference and another one that computes the `Union`, and then cleverly moving them around based on some flag. But that would mean more distance tests at runtime. Or maybe implementing a new `Curve` shape, that can be more flexible in its shape than the a circle. 
But both of these seem to me to be more complex and computationally expensive than having two shapes, but maybe I am missing something?